garage/tf/algos/ddpg.py

```diff
@@ -0,0 +1,376 @@
+    """
+    A DDPG model based on https://arxiv.org/pdf/1509.02971.pdf.
+
+    Example:
+        $ python garage/tf/launchers/ddpg_pendulum.py
+    """
+
+    def __init__(self,
+                 env,
+                 actor,
+                 critic,
+                 n_epochs=500,
+                 n_epoch_cycles=20,
+                 n_rollout_steps=100,
+                 n_train_steps=50,
+                 reward_scale=1.,
+                 batch_size=64,
+                 target_update_tau=0.01,
+                 discount=0.99,
+                 actor_lr=1e-4,
+                 critic_lr=1e-3,
+                 actor_weight_decay=0,
+                 critic_weight_decay=0,
+                 replay_buffer_size=int(1e6),
+                 min_buffer_size=10000,
+                 exploration_strategy=None,
+                 plot=False,
+                 pause_for_plot=False,
+                 actor_optimizer=None,
+                 critic_optimizer=None):
+        """
+        Construct class.
+
+        Args:
+            env(): Environment.
+            actor(garage.tf.policies.ContinuousMLPPolicy): Policy network.
+            critic(garage.tf.q_functions.ContinuousMLPQFunction):
+         Q Value network.
+            n_epochs(int, optional): Number of epochs.
+            n_epoch_cycles(int, optional): Number of epoch cycles.
+            n_rollout_steps(int, optional): Number of rollout steps.
+            n_train_steps(int, optional): Number of train steps.
+            reward_scale(float): The scaling factor applied to the rewards when
+         training.
+            batch_size(int): Number of samples for each minibatch.
+            target_update_tau(float): Interpolation parameter for doing the
+         soft target update.
+            discount(float): Discount factor for the cumulative return.
+            actor_lr(float): Learning rate for training policy network.
+            critic_lr(float): Learning rate for training q value network.
+            actor_weight_decay(float): L2 weight decay factor for parameters of
+         the policy network.
+            critic_weight_decay(float): L2 weight decay factor for parameters
+         of the q value network.
+            replay_buffer_size(int): Size of the replay buffer.
+            min_buffer_size(int): Minimum size of the replay buffer to start
+         training.
+            exploration_strategy(): Exploration strategy.
+            plot(bool): Whether to visualize the policy performance after each
+         eval_interval.
+            pause_for_plot(bool): Whether to pause before continuing when
+         plotting.
+            actor_optimizer(): Optimizer for training policy network.
+            critic_optimizer(): Optimizer for training q function network.
+        """
+        self.env = env
+
+        self.observation_dim = flat_dim(env.observation_space)
+        self.action_dim = flat_dim(env.action_space)
+        self.action_bound = env.action_space.high
```

Comment:
# from garage.envs.util import flat_dim
You can consider calling `_, self.action_bounds = bounds(env.action_space)`

Reply:
Thanks! Done.